File: dotnet/src/Core/Services/KernelStore.cs
using Microsoft.SemanticKernel;
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Agience.Core.Services
{
    public interface IKernelStore
    {
        Kernel GetKernel(string agentId);
        void AddKernel(string agentId, Kernel kernel);
        void RemoveKernel(string agentId);
    }

    public class KernelStore : IKernelStore
    {
        private readonly ConcurrentDictionary<string, Kernel> _kernels = new();

        public Kernel GetKernel(string agentId)
        {
            if (_kernels.TryGetValue(agentId, out var kernel))
            {
                return kernel;
            }

            throw new KeyNotFoundException($"Kernel not found for agent ID: {agentId}");
        }

        public void AddKernel(string agentId, Kernel kernel)
        {
            _kernels[agentId] = kernel;
        }

        public void RemoveKernel(string agentId)
        {
            _kernels.TryRemove(agentId, out _);
        }
    }

}

File: python/src/core/services/kernel_store.py
from typing import Dict
from semantic_kernel import Kernel
from threading import Lock


class IKernelStore:
    def get_kernel(self, agent_id: str) -> Kernel:
        pass

    def add_kernel(self, agent_id: str, kernel: Kernel) -> None:
        pass

    def remove_kernel(self, agent_id: str) -> None:
        pass


class KernelStore(IKernelStore):
    def __init__(self):
        self._kernels: Dict[str, Kernel] = {}
        self._lock = Lock()  # For thread safety

    def get_kernel(self, agent_id: str) -> Kernel:
        with self._lock:
            if agent_id in self._kernels:
                return self._kernels[agent_id]
            raise KeyError(f"Kernel not found for agent ID: {agent_id}")

    def add_kernel(self, agent_id: str, kernel: Kernel) -> None:
        with self._lock:
            self._kernels[agent_id] = kernel

    def remove_kernel(self, agent_id: str) -> None:
        with self._lock:
            self._kernels.pop(agent_id, None)
